libV2/vault/index.js

```diff
@@ -73,42 +73,87 @@ async function authorizeAccountAccessKey(authInfo, level, resources, log) {
     return [authed, authedRes];
 }
 
-// stub function to handle user keys until implemented
-async function authorizeUserAccessKey(authInfo, level, resources, authedRes, log) {
-    const authorizedResources = authedRes.reduce(
-        (authed, result) => {
-            if (result.isAllowed) {
-                // result.arn should be of format:
-                // arn:scality:utapi:::resourcetype/resource
-                assert(typeof result.arn === 'string');
-                assert(result.arn.indexOf('/') > -1);
-                const resource = result.arn.split('/')[1];
-                authed.push(resource);
-                log.trace('access granted for resource', { resource });
-            }
-            return authed;
-        }, [],
-    );
-
-    return [authedRes.length !== 0, authorizedResources];
+async function authorizeUserAccessKey(authInfo, level, resources, log) {
+    let authed = false;
+    let authedRes = [];
+
+    log.trace('Authorizing IAM user', { resources });
+
+    // Get the parent account id from the user's arn
+    const parentAccountId = authInfo.getArn().split(':')[4];
+
+    // All users require an attached policy to query metrics
+    // Additional filtering is performed here to limit access to the user's account
+    switch (level) {
+    // User keys can only query metrics their own account metrics
+    // So we can short circuit the auth to ->
+    // Did they request their account? Then authorize ONLY their account
+    case 'accounts': {
+        authed = resources.some(r => r === parentAccountId);
+        authedRes = authed ? [{ resource: parentAccountId, id: authInfo.getCanonicalID() }] : [];
+        break;
+    }
+
+    // Users can query other user's metrics if they are under the same account
```

Comment:
By default users have access to nothing. You need to attach a user policy to grant using the usual `<type>/<resource>` to give access. This check is so that a policy with `users/*` does not allow access to different accounts.